[json]: - added Player.SetViewMode/GetViewmode method to the json interface

Memphiz opened the pull request:
this bumps json api version to 10.2.0. For this I need help on how to handle the json api bump.
@MartijnKaijser  as reviewer because I hope he knows who else I can ask :D

## Description
expose the RenderManager.SetViewMode/GetViewMode method to json

## Motivation and Context
This was requested in the forum and is useful for projector users. When having a projector with 16:9 screen and watching 21:9 movies it is good to shift the whole videos own to the bottom of the screen to get a hard cut in contrast (black bars are usually not that black when projected). By exposing ViewMode to json it is now possible to set vertical shift to the liking.
I use the same approach for my projector too.
On Wetek I used the amlogic hardware feature to change the render rect. But now that I have switched to intel hardware I have to achieve the same thing via json now.

## How Has This Been Tested?
Tested via curl:

shift to the bottom ( vertical shift 1.0 means shift down as far as possible without cut):
curl -s --data-binary '{"jsonrpc":"2.0","method":"Player.SetViewMode","params":{"viewmode": {"verticalshift":1.0}},"id":1}' -H 'content-type: application/json;' http://localhost:8080/jsonrpc

back to normal view mode 
curl -s --data-binary '{"jsonrpc":"2.0","method":"Player.SetViewMode","params":{"viewmode": "normal"},"id":1}' -H 'content-type: application/json;' http://localhost:8080/jsonrpc

query current viewmode
curl -s --data-binary '{"jsonrpc":"2.0","method":"Player.GetViewMode","id":1}' -H 'content-type: application/json;' http://localhost:8080/jsonrpc

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Rechi commented on xbmc/interfaces/json-rpc/PlayerOperations.cpp:
```diff
@@ -489,6 +489,30 @@ JSONRPC_STATUS CPlayerOperations::Zoom(const std::string &method, ITransportLaye
+      verticalShift.isDouble() &&
+      stretch.isBoolean())
+  {
+    g_application.GetAppPlayer().SetRenderViewMode((int)viewMode.asInteger(),
```
use C++ style cast
```suggestion
    g_application.GetAppPlayer().SetRenderViewMode(static_cast<int>(viewMode.asInteger()),
```

Rechi commented on xbmc/interfaces/json-rpc/PlayerOperations.cpp:
```diff
@@ -489,6 +489,30 @@ JSONRPC_STATUS CPlayerOperations::Zoom(const std::string &method, ITransportLaye
+      stretch.isBoolean())
+  {
+    g_application.GetAppPlayer().SetRenderViewMode((int)viewMode.asInteger(),
+                                                   (double)zoom.asDouble(),
```
`asDouble` returns a `double`, no need to cast
```suggestion
                                                   static_cast<double>(zoom.asDouble()),
```

Rechi commented on xbmc/interfaces/json-rpc/PlayerOperations.cpp:
```diff
@@ -489,6 +489,30 @@ JSONRPC_STATUS CPlayerOperations::Zoom(const std::string &method, ITransportLaye
+  {
+    g_application.GetAppPlayer().SetRenderViewMode((int)viewMode.asInteger(),
+                                                   (double)zoom.asDouble(),
+                                                   (double)pixelRatio.asDouble(),
```
```suggestion
                                                   pixelRatio.asDouble(),
```

Rechi commented on xbmc/interfaces/json-rpc/PlayerOperations.cpp:
```diff
@@ -489,6 +489,30 @@ JSONRPC_STATUS CPlayerOperations::Zoom(const std::string &method, ITransportLaye
+    g_application.GetAppPlayer().SetRenderViewMode((int)viewMode.asInteger(),
+                                                   (double)zoom.asDouble(),
+                                                   (double)pixelRatio.asDouble(),
+                                                   (double)verticalShift.asDouble(),
```
```suggestion
                                                   verticalShift.asDouble(),
```

Rechi commented on xbmc/interfaces/json-rpc/PlayerOperations.cpp:
```diff
@@ -489,6 +489,30 @@ JSONRPC_STATUS CPlayerOperations::Zoom(const std::string &method, ITransportLaye
+                                                   (double)zoom.asDouble(),
+                                                   (double)pixelRatio.asDouble(),
+                                                   (double)verticalShift.asDouble(),
+                                                   (bool)stretch.asBoolean());
```
`asBoolean` returns a `bool`, no need to cast
```suggestion
                                                   stretch.asBoolean());
```

Memphiz commented:
@Rechi adapted

MartijnKaijser commented:
@DaveTBlake can you do sanity check on json stuff?

Memphiz commented:
Especially if the wiki is generated from the Schema or do I need to add the new method manually - and if the version bump is correct for an added method.

rmrector commented on xbmc/interfaces/json-rpc/schema/methods.json:
```diff
@@ -335,6 +335,20 @@
+      { "name": "zoom", "type": "number", "required": "false", "default": 1.0, "description": "Zoom were 1.0 means 100%" },
+      { "name": "pixelratio", "type": "number", "required": "false", "default": 1.0, "description": "Pixel aspect ratio were 1.0 means square pixel" },
+      { "name": "verticalshift", "type": "number", "required": "false", "default": 0.0, "description": "Verticalshift 1.0 means shift to bottom" },
+      { "name": "stretch", "type": "boolean", "required": "false", "default": false, "description": "Flag to enable nonlinear stretch" }
```
Can you change "stretch" to "nonlinearstretch" or "nonlinear"? Increasing the pixel ratio will linear-stretch the video so it would be nice to distinguish that in the name, and this external API is the place to break from internal API for clarity.

Memphiz replied:
done - changed to nonlinearstretch

rmrector commented:
The wiki is not generated from schema, and it's quite out of date already. FWIW I always prefer browsing the live API with something like Chorus2's [The Lab](https://github.com/xbmc/chorus2#kodi-api-browser).

The version bump is good.

Memphiz commented:
I will add the new method to the wiki once this PR is merged disregarding that it is outdated already :) ...

Memphiz commented:
The Lab - nice to know ... thats a huge tool :)

DaveTBlake commented:
@Memphiz thanks for doing some JSON. 
The version bump is correct (adding a method which is a non-breaking change).

I don't think there is a wiki page for v10 JSON schema yet. One was started for v9 by mistake (the odd versions are interim during dev phase) by mistake, and some things edited but most not. I think that that a partially updated and inaccurate page is worse than none at all, so unless someone volunteers to write a complete and accurate wiki for v10 I suggest it better to not have one. The schema is self documenting (description method) so it is not as if there is nothing for clients to look at.

But I do have some questions about the method
- If  there is a method to set viewmode then should there not also be a method to read the current viewmode?
- Are the "viewmode" integers codified somewhere?
- Do the double parameters have min and  max values? E.g. negative zoom, so will any value be OK?

Memphiz commented:
The viewmode has an enum 
https://github.com/xbmc/xbmc/blob/1b7e29a470492cde06fadc173ad349339993311a/xbmc/cores/VideoSettings.h#L68

But I struggled to duplicate it here because it would need adaptions to Json whenever the enum changed. It’s almost guaranteed to divert.

I can look for min max values for the double values though and add them to the Schema ...

Memphiz commented:
It add - if viewmode is a value which is not in the enum it will simply fallback to „normal“ viewmode. The inner code guarantees that.

Memphiz committed:
[JSON]Add Player.SetViewMode and Player.GetViewMode methods

Memphiz commented:
@DaveTBlake I tried my best. Have adde GetViewMode in addition and used reference types which are nullable. Whenever a parameter is not given the former setting is reused.

Memphiz commented:
This is annoying ...

DaveTBlake commented:
Thanks for this @Memphiz definitely better, and sorry for delay getting back to you (been busy in real life).  But I think it could be improved further. Those using the API should not have to look at C++ to know what the viewmode codes mean, the API should fully document and validate all parameters.  Yes this means keeping the JSON schema upto date with any enum changes in core code, but that is how API consumers know something has changed. 

Also looking deeper only some parameter combinations are valid. Calls to SetRenderViewMode either set viewmode to one of the mode presets e.g. Normal, Zoom, Stretch4x3 etc.  or set view mode to Custom and set one or more of zoom, pixelratio, verticalshift or nonlinearstretch. This can be implemented in JSON too.

Define Player.SetViewMode
```
"Player.SetViewMode": {
    "type": "method",
    "description": "Set view mode of video player",
    "transport": "Response",
    "permission": "ControlPlayback",
    "params": [
      { "name": "viewmode", "type": [
          { "$ref": "Player.CustomViewMode", "description": "Custom view mode", "required": true },
          { "name": "value", "$ref": "Player.ViewMode", "required": true}
        ],
      "required": true }
    ],
    "returns": "string"
  },
```
using new types added to https://github.com/xbmc/xbmc/blob/master/xbmc/interfaces/json-rpc/schema/types.json with
```
  "Player.ViewMode": {
    "type": "string",
    "enum": [  "normal", "zoom", "stretch4x3", "widezoom", "stretch16x9", "original",
               "stretch16x9nonlin", "zoom120width", "zoom110width" ]
  },
  "Player.CustomViewMode": {
    "type": "object",
    "required": true,
    "properties": {
      "zoom": { "$ref": "Optional.Number", "minimum":0.5, "maximum": 2.0, "description": "Zoom were 1.0 means 100%" },
      "pixelratio": { "$ref": "Optional.Number", "minimum":0.5, "maximum": 2.0, "description": "Pixel aspect ratio were 1.0 means square pixel" },
      "verticalshift": { "$ref": "Optional.Number", "minimum": -2.0, "maximum": 2.0, "description": "Verticalshift 1.0 means shift to bottom" },
      "nonlinearstretch": { "$ref": "Optional.Boolean", "description": "Flag to enable nonlinear stretch" }
    }
  },
```

Example valid calls then look like
```
Normal mode
{"id":1,"jsonrpc":"2.0","method":"Player.SetViewMode","params":{"viewmode": "normal"}}
Zoom 110 width
{"id":1,"jsonrpc":"2.0","method":"Player.SetViewMode","params":{"viewmode": "zoom110width"}}
Custom with zoom and vertical shift
{"id":1,"jsonrpc":"2.0","method":"Player.SetViewMode","params":{"viewmode": {"zoom": 1.6, "verticalshift": 0.5}}}
Custom with pixel ratio
{"id":1,"jsonrpc":"2.0","method":"Player.SetViewMode","params":{"viewmode": {"pixelratio": 1.01}}}
```
I pulled the max/min values for zoom, pixelratio and verticalshift from https://github.com/xbmc/xbmc/blob/master/xbmc/video/PlayerController.cpp, it makes sense to me they should be the same (assuming they exist for a reason?).  I can let you have a patch file for this if that helps.

DaveTBlake commented on xbmc/interfaces/json-rpc/PlayerOperations.cpp:
```diff
@@ -489,6 +490,61 @@ JSONRPC_STATUS CPlayerOperations::Zoom(const std::string &method, ITransportLaye
+  result["viewmode"] = g_application.GetAppPlayer().GetVideoSettings().m_ViewMode;
+  result["zoom"] = CDisplaySettings::GetInstance().GetZoomAmount();
+  result["pixelratio"] = CDisplaySettings::GetInstance().GetPixelRatio();
+  result["verticalshift"] = CDisplaySettings::GetInstance().GetVerticalShift();
```
Can nonlinearstretch boolean be returned too?

Memphiz replied:
No it’s internally used but not exposed by settings.

DaveTBlake replied:
How about `CDisplaySettings::GetInstance().IsNonLinearStretched()` ?

Memphiz commented:
So is our JSON API in that shape you suggest here everywhere else? This will take me a while to do it. I considered this an advanced feature but I guess everything in that API needs to be bullet proof.
Tbh - I start regretting doing this PR at all. My time is very limited atm. Looks like there are no quick win PRs common anymore nowadays.
Putting this on hold for now.

DaveTBlake commented:
Yeap, needs to be bullet proof even if  the internal function it exposes does not have the cleanest interface. Understand the time limits etc. @Memphiz, and sorry for any frustration. This can still be a quickish win. I have a patch file that implements what I'm suggesting, shall I push it to your repo (if I can)? Work all done, just a matter of getting into into this PR :)

Memphiz commented:
Sure or to my repo and i merge it into this PR - thx

DaveTBlake commented:
@Memphiz  I  have raised a PR for this against your repo, not sure if that is best way so if you need me to do something else then just let me know.

Memphiz commented:
@DaveTBlake i have added your commit to this PR and added my refactoring proposal. Main goal was to reduce the code duplication, early returns and if nesting. Also we have the string to enum mapping in a single place now which makes it easier to add new values later on. Once you are ok with this (and maybe you even have some time to test it) I will squash this into a single commit.

DaveTBlake commented:
@Memphiz I'm fine in principle with the refactoring, I obviously don't have as much aversion to early returns as you do and tend to copy existing code style :)

Only thought looking at the code is that `IsValidViewmodeParameter()` isn't needed as the JSON schema definition ensures the parameters are valid in that way. A request with something other than "increase"/"decrease" or a double would be rejected before SetViewMode is called. I can see how you would derive it from my code,  it was my mistake too.

DaveTBlake commented on xbmc/interfaces/json-rpc/PlayerOperations.cpp:
```diff
@@ -489,6 +492,150 @@ JSONRPC_STATUS CPlayerOperations::Zoom(const std::string &method, ITransportLaye
+  return result;
+}
+
+bool IsValidViewmodeParameter(const CVariant &parameter)
```
This basic parameter validation is performed by the JSON schema definition, if parameters are invalid in this way the request is returned with an error before SetViewMode is called

Memphiz replied:
removed the checks

DaveTBlake commented on xbmc/interfaces/json-rpc/PlayerOperations.cpp:
```diff
@@ -489,6 +492,150 @@ JSONRPC_STATUS CPlayerOperations::Zoom(const std::string &method, ITransportLaye
   }
 }
 
+std::map<std::string, ViewMode> viewModes =
```
Thanks, good to have clear mapping of the relation between the two enum.
Could you add comment
// Matching pairs of  values from JSON type "Player.ViewMode" and C++ enum ViewMode
// Additions to enum ViewMode need to be added here and in the JSON type

Memphiz replied:
done

Memphiz commented:
updated and squashed @DaveTBlake

DaveTBlake commented:
Great @Memphiz, thanks for this.
I have updated the test call example in the first post, and I will do some more through testing of all the variations later but looks good.

DaveTBlake reviewed (approved):
Done extensive testing and all looking good. 
This is ready to merge as far as I'm concerned, and a useful JSON improvement for 18.2

DaveTBlake commented:
jenkins build this please

DaveTBlake commented:
jenkins build this please
